// @parcel/bundler-default/lib/DefaultBundler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
function _graph() {
  const data = require("@parcel/graph");
  _graph = function () {
    return data;
  };
  return data;
}
function _assert() {
  const data = _interopRequireDefault(require("assert"));
  _assert = function () {
    return data;
  };
  return data;
}
function _plugin() {
  const data = require("@parcel/plugin");
  _plugin = function () {
    return data;
  };
  return data;
}
function _utils() {
  const data = require("@parcel/utils");
  _utils = function () {
    return data;
  };
  return data;
}
function _nullthrows() {
  const data = _interopRequireDefault(require("nullthrows"));
  _nullthrows = function () {
    return data;
  };
  return data;
}
function _diagnostic() {
  const data = require("@parcel/diagnostic");
  _diagnostic = function () {
    return data;
  };
  return data;
}
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* eslint-disable */ // Default options by http version.
const HTTP_OPTIONS = {
  '1': {
    minBundles: 1,
    minBundleSize: 30000,
    maxParallelRequests: 6
  },
  '2': {
    minBundles: 1,
    minBundleSize: 20000,
    maxParallelRequests: 25
  }
};

/* BundleRoot - An asset that is the main entry of a Bundle. */

const dependencyPriorityEdges = {
  sync: 1,
  parallel: 2,
  lazy: 3
};

// IdealGraph is the structure we will pass to decorate,
// which mutates the assetGraph into the bundleGraph we would
// expect from default bundler
/**
 *
 * The Bundler works by creating an IdealGraph, which contains a BundleGraph that models bundles
 * connected to other bundles by what references them, and thus models BundleGroups.
 *
 * First, we enter `bundle({bundleGraph, config})`. Here, "bundleGraph" is actually just the
 * assetGraph turned into a type `MutableBundleGraph`, which will then be mutated in decorate,
 * and turned into what we expect the bundleGraph to be as per the old (default) bundler structure
 *  & what the rest of Parcel expects a BundleGraph to be.
 *
 * `bundle({bundleGraph, config})` First gets a Mapping of target to entries, In most cases there is
 *  only one target, and one or more entries. (Targets are pertinent in monorepos or projects where you
 *  will have two or more distDirs, or output folders.) Then calls create IdealGraph and Decorate per target.
 *
 */
var _default = new (_plugin().Bundler)({
  loadConfig({
    config,
    options
  }) {
    return loadBundlerConfig(config, options);
  },
  bundle({
    bundleGraph,
    config
  }) {
    let targetMap = getEntryByTarget(bundleGraph); // Organize entries by target output folder/ distDir
    let graphs = [];
    for (let entries of targetMap.values()) {
      // Create separate bundleGraphs per distDir
      graphs.push(createIdealGraph(bundleGraph, config, entries));
    }
    for (let g of graphs) {
      decorateLegacyGraph(g, bundleGraph); //mutate original graph
    }
  },

  optimize() {}
});
exports.default = _default;
function decorateLegacyGraph(idealGraph, bundleGraph) {
  let idealBundleToLegacyBundle = new Map();
  let {
    bundleGraph: idealBundleGraph,
    dependencyBundleGraph,
    bundleGroupBundleIds
  } = idealGraph;
  let entryBundleToBundleGroup = new Map();
  // Step Create Bundles: Create bundle groups, bundles, and shared bundles and add assets to them
  for (let [bundleNodeId, idealBundle] of idealBundleGraph.nodes) {
    if (idealBundle === 'root') continue;
    let entryAsset = idealBundle.mainEntryAsset;
    let bundleGroup;
    let bundle;
    if (bundleGroupBundleIds.has(bundleNodeId)) {
      let dependencies = dependencyBundleGraph.getNodeIdsConnectedTo(dependencyBundleGraph.getNodeIdByContentKey(String(bundleNodeId)), _graph().ALL_EDGE_TYPES).map(nodeId => {
        let dependency = (0, _nullthrows().default)(dependencyBundleGraph.getNode(nodeId));
        (0, _assert().default)(dependency.type === 'dependency');
        return dependency.value;
      });
      for (let dependency of dependencies) {
        bundleGroup = bundleGraph.createBundleGroup(dependency, idealBundle.target);
      }
      (0, _assert().default)(bundleGroup);
      entryBundleToBundleGroup.set(bundleNodeId, bundleGroup);
      bundle = (0, _nullthrows().default)(bundleGraph.createBundle({
        entryAsset: (0, _nullthrows().default)(entryAsset),
        needsStableName: idealBundle.needsStableName,
        bundleBehavior: idealBundle.bundleBehavior,
        target: idealBundle.target
      }));
      bundleGraph.addBundleToBundleGroup(bundle, bundleGroup);
    } else if (idealBundle.sourceBundles.size > 0 && !idealBundle.mainEntryAsset) {
      bundle = (0, _nullthrows().default)(bundleGraph.createBundle({
        uniqueKey: [...idealBundle.assets].map(asset => asset.id).join(',') + [...idealBundle.sourceBundles].join(','),
        needsStableName: idealBundle.needsStableName,
        bundleBehavior: idealBundle.bundleBehavior,
        type: idealBundle.type,
        target: idealBundle.target,
        env: idealBundle.env
      }));
    } else if (idealBundle.uniqueKey != null) {
      bundle = (0, _nullthrows().default)(bundleGraph.createBundle({
        uniqueKey: idealBundle.uniqueKey,
        needsStableName: idealBundle.needsStableName,
        bundleBehavior: idealBundle.bundleBehavior,
        type: idealBundle.type,
        target: idealBundle.target,
        env: idealBundle.env
      }));
    } else {
      (0, _assert().default)(entryAsset != null);
      bundle = (0, _nullthrows().default)(bundleGraph.createBundle({
        entryAsset,
        needsStableName: idealBundle.needsStableName,
        bundleBehavior: idealBundle.bundleBehavior,
        target: idealBundle.target
      }));
    }
    idealBundleToLegacyBundle.set(idealBundle, bundle);
    for (let asset of idealBundle.assets) {
      bundleGraph.addAssetToBundle(asset, bundle);
    }
  }
  // Step Internalization: Internalize dependencies for bundles
  for (let [, idealBundle] of idealBundleGraph.nodes) {
    if (idealBundle === 'root') continue;
    let bundle = (0, _nullthrows().default)(idealBundleToLegacyBundle.get(idealBundle));
    if (idealBundle.internalizedAssets) {
      for (let internalized of idealBundle.internalizedAssets.values()) {
        let incomingDeps = bundleGraph.getIncomingDependencies(internalized);
        for (let incomingDep of incomingDeps) {
          if (incomingDep.priority === 'lazy' && incomingDep.specifierType !== 'url' && bundle.hasDependency(incomingDep)) {
            bundleGraph.internalizeAsyncDependency(bundle, incomingDep);
          }
        }
      }
    }
  }

  // Step Add to BundleGroups: Add bundles to their bundle groups
  idealBundleGraph.traverse((nodeId, _, actions) => {
    let node = idealBundleGraph.getNode(nodeId);
    if (node === 'root') {
      return;
    }
    actions.skipChildren();
    let outboundNodeIds = idealBundleGraph.getNodeIdsConnectedFrom(nodeId);
    let entryBundle = (0, _nullthrows().default)(idealBundleGraph.getNode(nodeId));
    (0, _assert().default)(entryBundle !== 'root');
    let legacyEntryBundle = (0, _nullthrows().default)(idealBundleToLegacyBundle.get(entryBundle));
    for (let id of outboundNodeIds) {
      let siblingBundle = (0, _nullthrows().default)(idealBundleGraph.getNode(id));
      (0, _assert().default)(siblingBundle !== 'root');
      let legacySiblingBundle = (0, _nullthrows().default)(idealBundleToLegacyBundle.get(siblingBundle));
      bundleGraph.createBundleReference(legacyEntryBundle, legacySiblingBundle);
    }
  });

  // Step References: Add references to all bundles
  for (let [asset, references] of idealGraph.assetReference) {
    for (let [dependency, bundle] of references) {
      let legacyBundle = (0, _nullthrows().default)(idealBundleToLegacyBundle.get(bundle));
      bundleGraph.createAssetReference(dependency, asset, legacyBundle);
    }
  }
  for (let {
    from,
    to
  } of idealBundleGraph.getAllEdges()) {
    let sourceBundle = (0, _nullthrows().default)(idealBundleGraph.getNode(from));
    if (sourceBundle === 'root') {
      continue;
    }
    (0, _assert().default)(sourceBundle !== 'root');
    let legacySourceBundle = (0, _nullthrows().default)(idealBundleToLegacyBundle.get(sourceBundle));
    let targetBundle = (0, _nullthrows().default)(idealBundleGraph.getNode(to));
    if (targetBundle === 'root') {
      continue;
    }
    (0, _assert().default)(targetBundle !== 'root');
    let legacyTargetBundle = (0, _nullthrows().default)(idealBundleToLegacyBundle.get(targetBundle));
    bundleGraph.createBundleReference(legacySourceBundle, legacyTargetBundle);
  }
}
function createIdealGraph(assetGraph, config, entries) {
  // Asset to the bundle and group it's an entry of
  let bundleRoots = new Map();
  let bundles = new Map();
  let dependencyBundleGraph = new (_graph().ContentGraph)();
  let assetReference = new (_utils().DefaultMap)(() => []);

  // A Graph of Bundles and a root node (dummy string), which models only Bundles, and connections to their
  // referencing Bundle. There are no actual BundleGroup nodes, just bundles that take on that role.
  let bundleGraph = new (_graph().Graph)();
  let stack = [];
  let bundleRootEdgeTypes = {
    parallel: 1,
    lazy: 2
  };
  // ContentGraph that models bundleRoots, with parallel & async deps only to inform reachability
  let bundleRootGraph = new (_graph().ContentGraph)();
  let bundleGroupBundleIds = new Set();

  // Models bundleRoots and the assets that require it synchronously
  let reachableRoots = new (_graph().ContentGraph)();
  let rootNodeId = (0, _nullthrows().default)(bundleRootGraph.addNode('root'));
  let bundleGraphRootNodeId = (0, _nullthrows().default)(bundleGraph.addNode('root'));
  bundleRootGraph.setRootNodeId(rootNodeId);
  bundleGraph.setRootNodeId(bundleGraphRootNodeId);
  // Step Create Entry Bundles
  for (let [asset, dependency] of entries) {
    let bundle = createBundle({
      asset,
      target: (0, _nullthrows().default)(dependency.target),
      needsStableName: dependency.isEntry
    });
    let nodeId = bundleGraph.addNode(bundle);
    bundles.set(asset.id, nodeId);
    bundleRoots.set(asset, [nodeId, nodeId]);
    bundleRootGraph.addEdge(rootNodeId, bundleRootGraph.addNodeByContentKey(asset.id, asset));
    bundleGraph.addEdge(bundleGraphRootNodeId, nodeId);
    dependencyBundleGraph.addEdge(dependencyBundleGraph.addNodeByContentKeyIfNeeded(dependency.id, {
      value: dependency,
      type: 'dependency'
    }), dependencyBundleGraph.addNodeByContentKeyIfNeeded(String(nodeId), {
      value: bundle,
      type: 'bundle'
    }), dependencyPriorityEdges[dependency.priority]);
    bundleGroupBundleIds.add(nodeId);
  }
  let assets = [];
  let typeChangeIds = new Set();
  /**
   * Step Create Bundles: Traverse the assetGraph (aka MutableBundleGraph) and create bundles
   * for asset type changes, parallel, inline, and async or lazy dependencies,
   * adding only that asset to each bundle, not its entire subgraph.
   */
  assetGraph.traverse({
    enter(node, context, actions) {
      if (node.type === 'asset') {
        if ((context === null || context === void 0 ? void 0 : context.type) === 'dependency' && context !== null && context !== void 0 && context.value.isEntry && !entries.has(node.value)) {
          // Skip whole subtrees of other targets by skipping those entries
          actions.skipChildren();
          return node;
        }
        assets.push(node.value);
        let bundleIdTuple = bundleRoots.get(node.value);
        if (bundleIdTuple && bundleIdTuple[0] === bundleIdTuple[1]) {
          // Push to the stack (only) when a new bundle is created
          stack.push([node.value, bundleIdTuple[0]]);
        } else if (bundleIdTuple) {
          // Otherwise, push on the last bundle that marks the start of a BundleGroup
          stack.push([node.value, stack[stack.length - 1][1]]);
        }
      } else if (node.type === 'dependency') {
        if (context == null) {
          return node;
        }
        let dependency = node.value;
        if (assetGraph.isDependencySkipped(dependency)) {
          actions.skipChildren();
          return node;
        }
        (0, _assert().default)((context === null || context === void 0 ? void 0 : context.type) === 'asset');
        let parentAsset = context.value;
        let assets = assetGraph.getDependencyAssets(dependency);
        if (assets.length === 0) {
          return node;
        }
        for (let childAsset of assets) {
          if (dependency.priority === 'lazy' || childAsset.bundleBehavior === 'isolated' // An isolated Dependency, or Bundle must contain all assets it needs to load.
          ) {
            let bundleId = bundles.get(childAsset.id);
            let bundle;
            if (bundleId == null) {
              var _dependency$bundleBeh;
              let firstBundleGroup = (0, _nullthrows().default)(bundleGraph.getNode(stack[0][1]));
              (0, _assert().default)(firstBundleGroup !== 'root');
              bundle = createBundle({
                asset: childAsset,
                target: firstBundleGroup.target,
                needsStableName: dependency.bundleBehavior === 'inline' || childAsset.bundleBehavior === 'inline' ? false : dependency.isEntry || dependency.needsStableName,
                bundleBehavior: (_dependency$bundleBeh = dependency.bundleBehavior) !== null && _dependency$bundleBeh !== void 0 ? _dependency$bundleBeh : childAsset.bundleBehavior
              });
              bundleId = bundleGraph.addNode(bundle);
              bundles.set(childAsset.id, bundleId);
              bundleRoots.set(childAsset, [bundleId, bundleId]);
              bundleGroupBundleIds.add(bundleId);
